### DISCUSSION TITLE
Use dynamic routes while maintaining existing file structure

### DISCUSSION THREAD

**Discussion by bfeliz** [vercel/next.js#17664] [Help] [answered]
I currently have a file structure like so within the 'src' directory:
```bash
|____pages
| |____FirstPage
| | |____First.js
| | |____index.js
| |____SecondPage
| | |____Second.js
| | |____index.js
| |____FirstDynamicRoutePage
| | |____PageToRenderOnDynamicRoute.js
| | |____index.js
```
I would like to keep this file structure and use dynamic routing. For example, the route for 'PageToRenderOnDynamicRoute.js" will be something like '/page/pageType/:pageId/otherSlug/:slug'.

I am aware that I can use the dynamic routing provided by Next.js, but I do not want to use the necessary file structure of:
```bash
| |____page
| | |____pageType
| | | |____[pageId]
| | | | |____otherSlug
| | | | | |____[slug].js
```
I have tried using [next-routes](https://github.com/fridays/next-routes), but ran into many issues, including: returned 404 when using getStaticProps or getServerSideProps, returned 404 on refresh of page, returned 404 when url typed in manually.

Does anyone know of a way that I can keep my existing file structure **AND** use dynamic routing? I am open to any options, including continuing to try next-routes if you are aware of working solutions to the aforementioned issues. I appreciate your help!

**Comment by francisrod01** [comment]
> Next.js has the file-system based router built on the concept of pages.
> 
> When a file is added to the `pages` directory it's automatically available as a route.
> The files inside the pages directory can be used to define most common patterns.

I encourage you to use that structure as components for pages.

**Reply by bfeliz** [reply]
Thank you for the reply, but I have been asked by my team to try and preserve our existing file structure if at all possible instead of using the nested file system.

**Comment by ijjk** [comment, marked as answer]
Hi, have you tried leveraging the built-in [rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites) feature. This should allow you to map pages with dynamic params without changing your file structure. On the latest canary of Next.js (soon to be stable) the `next/link` component has also been updated to handle these rewrites on the client more seamlessly.

**Reply by bfeliz** [reply]
Thank you for the quick reply! I have tested out the rewrites feature, but so far no luck, only 404 errors returned. I have updated to version 9.5.3, and this is how my next.config.js looks:
```js
const withSass = require('@zeit/next-sass')
const withTM = require('next-transpile-modules');
const withCSS = require("@zeit/next-css");
const withImages = require('next-images')

module.exports = withImages(withSass(withCSS(withTM({
  async rewrites() {
    return [
      { source: '/page/pageType/:slug*', destination: '/FirstDynamicRoutePage' },
    ];
  },
  transpileModules: ['react-bulma-components'],
  sassLoaderOptions: {
    includePaths: ["./src"]
  },
  module: {
    rules: [
      {
        test: /\.css$/i,
        use: ['css-loader'],
      },
    ],
  },
}))));
```

My link looks like:
```js
      <Link href={`/page/pageType/${props.pageId}/otherSlug/${props.slug}`}>
        <a>
          {props.name}
        </a>
      </Link>
```
We do have a custom server in place but I do not think that it should be affecting anything because if I use the dynamic routing file structure it works perfectly fine even with the custom server. Any thoughts? Am I somehow misusing the rewrites feature?

**Reply by ijjk** [reply]
Interesting, I'm not able to reproduce a 404 with that rewrite and page set-up so it might be from the custom server in front of it although it shouldn't affect it as long as all 404s in the custom server are being passed to Next.js' request handler. If you can provide a repo with a minimal reproduction where this isn't working properly I'd be happy to take a look!

**Reply by bfeliz** [reply]
So turns out my custom server was still using the next-routes handler, so I was able to replace that with the Next.js request handler, and also restructure the next.config.js like so for the rewrites to function correctly:
```js
const withSass = require('@zeit/next-sass')
const withTM = require('next-transpile-modules');
const withCSS = require("@zeit/next-css");
const withImages = require('next-images')

const myModule = module.exports = withImages(withSass(withCSS(withTM({
  transpileModules: ['react-bulma-components'],
  sassLoaderOptions: {
    includePaths: ["./src"]
  },
  module: {
    rules: [
      {
        test: /\.css$/i,
        use: ['css-loader'],
      },
    ],
  },
}))));

myModule.rewrites = async () => {
  return [
      { source: '/page/pageType/:slug*', destination: '/FirstDynamicRoutePage' },
  ];
};
```
Thank you for pointing me in the the right direction, @ijjk! I appreciate your help and your willingness to dig deeper even though it turned out your suggestions alone helped immensely :)
